### PR TITLE
MergeTree FINAL extract non intersecting parts ranges

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Extract non intersecting parts ranges from MergeTree table during FINAL processing. That way we can avoid additional FINAL logic for this non intersecting parts ranges. In case when amount of duplicate values with same primary key is low, performance will be almost the same as without FINAL. Improve reading performance for MergeTree FINAL when `do_not_merge_across_partitions_select_final` setting is set.

### CODE REVIEW

**Comment by kitaisreal** [comment]
Table without duplicates:

```
CREATE TABLE test_table (id UInt64, value String) ENGINE=ReplacingMergeTree ORDER BY id SETTINGS index_granularity = 8192;
INSERT INTO test_table SELECT number, number FROM numbers(500000000);

OPTIMIZE TABLE test_table;

SELECT name, part_type, marks FROM system.parts WHERE table = 'test_table' AND active = 1;

┌─name────────┬─part_type─┬─marks─┐
│ all_1_477_4 │ Wide      │ 61037 │
└─────────────┴───────────┴───────┘
```

Before:
```
SELECT * FROM test_table FINAL FORMAT Null

0 rows in set. Elapsed: 3.464 sec. Processed 500.12 million rows, 12.89 GB (144.36 million rows/s., 3.72 GB/s.)
Peak memory usage: 178.61 MiB.
```

After:
```
SELECT * FROM test_table FINAL FORMAT Null

0 rows in set. Elapsed: 0.525 sec. Processed 500.00 million rows, 12.89 GB (952.04 million rows/s., 24.54 GB/s.)
Peak memory usage: 43.50 MiB.
```

The same as without final:
```
SELECT * FROM test_table FORMAT Null

0 rows in set. Elapsed: 0.526 sec. Processed 500.00 million rows, 12.89 GB (950.44 million rows/s., 24.50 GB/s.)
Peak memory usage: 40.77 MiB.
```

Table with some intersecting ranges:

```
DROP TABLE IF EXISTS test_table_intersecting_ranges;
CREATE TABLE test_table_intersecting_ranges (id UInt64, value String) ENGINE=ReplacingMergeTree ORDER BY id SETTINGS index_granularity = 8192;
INSERT INTO test_table_intersecting_ranges SELECT number, number FROM numbers(250000000);
INSERT INTO test_table_intersecting_ranges SELECT number + 200000000, number FROM numbers(250000000);
SYSTEM STOP MERGES test_table_intersecting_ranges;
```

Before:
```
SELECT * FROM test_table_intersecting_ranges FINAL FORMAT Null

0 rows in set. Elapsed: 3.279 sec. Processed 500.16 million rows, 12.78 GB (152.53 million rows/s., 3.90 GB/s.)
Peak memory usage: 196.35 MiB.
```

After:
```
SELECT * FROM test_table_intersecting_ranges FINAL FORMAT Null

0 rows in set. Elapsed: 1.379 sec. Processed 500.06 million rows, 12.78 GB (362.61 million rows/s., 9.27 GB/s.)
Peak memory usage: 128.50 MiB.
```

Setting do_not_merge_across_partitions_select_final performance improvements:

Create table:
```
DROP TABLE IF EXISTS test_table_partitioned;
CREATE TABLE test_table_partitioned (id UInt64, value String) ENGINE=ReplacingMergeTree ORDER BY id PARTITION BY id % 10 SETTINGS index_granularity = 8192;
INSERT INTO test_table_partitioned SELECT number, number FROM numbers(500000000);
OPTIMIZE TABLE test_table_partitioned;

SELECT name, part_type, marks FROM system.parts WHERE table = 'test_table_partitioned' AND active = 1;

┌─name────────┬─part_type─┬─marks─┐
│ 0_1_4767_5  │ Compact   │  6105 │
│ 1_2_4768_5  │ Compact   │  6105 │
│ 2_3_4769_5  │ Compact   │  6105 │
│ 3_4_4770_5  │ Compact   │  6105 │
│ 4_5_4761_5  │ Compact   │  6105 │
│ 5_6_4762_5  │ Compact   │  6105 │
│ 6_7_4763_5  │ Compact   │  6105 │
│ 7_8_4764_5  │ Compact   │  6105 │
│ 8_9_4765_5  │ Compact   │  6105 │
│ 9_10_4766_5 │ Compact   │  6105 │
└─────────────┴───────────┴───────┘
```

Before:

```
SELECT * FROM test_table_partitioned FINAL SETTINGS do_not_merge_across_partitions_select_final = 1 FORMAT Null

0 rows in set. Elapsed: 0.891 sec. Processed 500.00 million rows, 12.89 GB (560.99 million rows/s., 14.46 GB/s.)
Peak memory usage: 32.01 MiB.
```

After:

```
SELECT * FROM test_table_partitioned FINAL SETTINGS do_not_merge_across_partitions_select_final = 1 FORMAT Null

0 rows in set. Elapsed: 0.610 sec. Processed 500.00 million rows, 12.89 GB (819.56 million rows/s., 21.13 GB/s.)
Peak memory usage: 54.96 MiB.
```

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit f83578bccfe1c30ae0c52ae8f7a6c18fec6f4954 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58120/f83578bccfe1c30ae0c52ae8f7a6c18fec6f4954/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58120/f83578bccfe1c30ae0c52ae8f7a6c18fec6f4954/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by kitaisreal** [comment]
> For ReplacingMergeTree, test like that is more likely
> 
> ```
> DROP TABLE IF EXISTS test_table_intersecting_ranges;
> CREATE TABLE test_table_intersecting_ranges (id UInt64, value String) ENGINE=ReplacingMergeTree ORDER BY id SETTINGS index_granularity = 8192;
> INSERT INTO test_table_intersecting_ranges SELECT number , number FROM numbers(500000000);
> INSERT INTO test_table_intersecting_ranges SELECT number*50000, number FROM numbers(20);
> 
> SYSTEM STOP MERGES test_table_intersecting_ranges;
> ```
Thanks for test.

Before:
```
SELECT * FROM test_table_intersecting_ranges FINAL FORMAT Null

0 rows in set. Elapsed: 3.343 sec. Processed 500.12 million rows, 12.89 GB (149.62 million rows/s., 3.86 GB/s.)
Peak memory usage: 239.85 MiB.
```

After:
```
SELECT * FROM test_table_intersecting_ranges FINAL FORMAT Null

0 rows in set. Elapsed: 0.629 sec. Processed 489.72 million rows, 12.62 GB (779.03 million rows/s., 20.08 GB/s.)
Peak memory usage: 46.30 MiB.
```

**Comment by kitaisreal** [comment]
Performance results.

X86-64:
https://s3.amazonaws.com/clickhouse-test-reports/58120/a85e2cd6708ca3f6d6ce97e2243db85848b8047e/performance_comparison_[1_4]/report.html
https://s3.amazonaws.com/clickhouse-test-reports/58120/a85e2cd6708ca3f6d6ce97e2243db85848b8047e/performance_comparison_[4_4]/report.html

ARM:
https://s3.amazonaws.com/clickhouse-test-reports/58120/a85e2cd6708ca3f6d6ce97e2243db85848b8047e/performance_comparison_aarch64_[1_4]/report.html
https://s3.amazonaws.com/clickhouse-test-reports/58120/a85e2cd6708ca3f6d6ce97e2243db85848b8047e/performance_comparison_aarch64_[4_4]/report.html

**Review by alexey-milovidov — APPROVED** [review]

File: src/Processors/QueryPlan/PartsSplitter.cpp [hunk 1]
```diff
@@ -49,81 +65,451 @@ class IndexAccess
...
+      *
+      * range 1    [    ...              ----]
+      * range 2    [    ...    (value_1)]
+      *
+      * In this case we can extract non intersecting part of range 1. This non intersecting part will have start
+      * of leftmost mark from range 1 that contains value greater than value_1 and end with range 1 end.
+      *
+      * Additional details:
+      *
+      * 1. If part level is 0, we must process all ranges from this part, because they can contain duplicate primary keys.
```
Comment (tavplubix) [inline src/Processors/QueryPlan/PartsSplitter.cpp:286]: This assumption is incorrect for `ReplacingMergeTree` with `is_deleted` column, and that's why `02814_ReplacingMergeTree_fix_select_final_on_single_partition` was broken
See also https://github.com/ClickHouse/ClickHouse/commit/95e4b0002f2f5c64581268c341ee895c426e1a47

**Comment by tavplubix** [comment]
The code is quite tricky and it handles many corner cases, so please consider introducing a setting like `optimize_non_intersecting_ranges_on_final` that allows disabling this optimization (just in case).

Also, there's a lack of introspection/logging, so it will be hard to debug potential issues

**Pull request merged**